packages/react-instantsearch-hooks/src/components/__tests__/InstantSearch.test.tsx

```diff
@@ -508,11 +510,244 @@ describe('InstantSearch', () => {
+
+    function App({ searchClient }) {
+      return (
+        <StrictMode>
+          <InstantSearch searchClient={searchClient} indexName="indexName">
+            <SearchBox />
+          </InstantSearch>
+        </StrictMode>
+      );
+    }
+
+    const { rerender } = render(<App searchClient={searchClient1} />);
+
+    await waitFor(() => {
+      expect(searchClient1.search).toHaveBeenCalledTimes(1);
+    });
+
+    rerender(<App searchClient={searchClient2} />);
+
+    await waitFor(() => {
+      expect(searchClient1.search).toHaveBeenCalledTimes(1);
+      // @TODO: can this call count go down to 1 in Strict Mode?
+      expect(searchClient2.search).toHaveBeenCalledTimes(2);
+      expect(searchClient2.addAlgoliaAgent).toHaveBeenCalledWith(
+        `react (${ReactVersion})`
+      );
+      expect(searchClient2.addAlgoliaAgent).toHaveBeenCalledWith(
+        `react-instantsearch (${version})`
+      );
+      expect(searchClient2.addAlgoliaAgent).toHaveBeenCalledWith(
+        `react-instantsearch-hooks (${version})`
+      );
+    });
+
+    rerender(<App searchClient={searchClient3} />);
+
+    await waitFor(() => {
+      expect(searchClient1.search).toHaveBeenCalledTimes(1);
+      expect(searchClient2.search).toHaveBeenCalledTimes(2);
+      expect(searchClient3.search).toHaveBeenCalledTimes(2);
+      expect(searchClient3.addAlgoliaAgent).toHaveBeenCalledWith(
+        `react (${ReactVersion})`
+      );
+      expect(searchClient3.addAlgoliaAgent).toHaveBeenCalledWith(
+        `react-instantsearch (${version})`
+      );
+      expect(searchClient3.addAlgoliaAgent).toHaveBeenCalledWith(
+        `react-instantsearch-hooks (${version})`
+      );
+    });
+
+    userEvent.type(screen.getByRole('searchbox'), 'iphone');
+
+    await waitFor(() => {
+      expect(searchClient3.search).toHaveBeenCalledTimes(8);
+      expect(searchClient3.search).toHaveBeenLastCalledWith([
+        {
+          indexName: 'indexName',
+          params: expect.objectContaining({
+            query: 'iphone',
+          }),
+        },
+      ]);
+    });
+  });
+
+  test('updates the index on index prop change', async () => {
+    const searchClient = createSearchClient({});
+
+    function App({ indexName }) {
+      return (
+        <StrictMode>
+          <InstantSearch searchClient={searchClient} indexName={indexName}>
+            <SearchBox />
+          </InstantSearch>
+        </StrictMode>
+      );
+    }
+
+    const { rerender } = render(<App indexName="indexName1" />);
+
+    await waitFor(() => {
+      expect(searchClient.search).toHaveBeenCalledTimes(1);
+      expect(searchClient.search).toHaveBeenLastCalledWith([
+        expect.objectContaining({
+          indexName: 'indexName1',
+        }),
+      ]);
+    });
+
+    rerender(<App indexName="indexName2" />);
+
+    await waitFor(() => {
+      // @TODO: can this call count go down in Strict Mode?
+      expect(searchClient.search).toHaveBeenCalledTimes(3);
```

Comment:
do we know why one extra search happened with index 1? is it the same parameters?

Reply:
These are the `searchClient.search()` mock calls at this step:

```json
[
      [
        [
          {
            "indexName": "indexName1",
            "params": {
              "facets": [],
              "query": "",
              "tagFilters": ""
            }
          }
        ]
      ],
      [
        [
          {
            "indexName": "indexName2",
            "params": {
              "facets": [],
              "query": "",
              "tagFilters": ""
            }
          }
        ]
      ],
      [
        [
          {
            "indexName": "indexName2",
            "params": {
              "facets": [],
              "query": "",
              "tagFilters": ""
            }
          }
        ]
      ]
    ]
```

So at least the client is able to cache it, but I'm not yet sure why there's this extra request.

Reply:
Storing the prev props in a ref rather than in state triggers the correct amount of searches. See 2bb00c6220dae6ceb9a05615ff255840f8d38214.